File: 0div-idbig-omega-company-tagsdiv-idbig-omega-topbardiv-classcompanytagscontainer-styleoverflow-x-scroll-flex-wrap-nowrap-div-classcompanytagscontainer-tag-stylebackground-color-rgba0-10-32-005-divfacebook-divdiv-classcompanytagscontainer-tagoccurence3-div-div-divdiv-classcompanytagscontainer-chevrondivsvg-version11-idicon-xmlnshttp-wwww3org-2000-svg-xmlns-xlinkhttp-wwww3org-1999-xlink-x0px-y0px-viewbox0-0-32-32-fill4087f1-xml-spacepreserve-stylewidth-20px-polygon-points16-22-6-12-74-106-16-192-246-106-26-12-polygonrect-id-x3c-transparent-rectangle-x3e-classst0-fillnone-width32-height32-rect-svg-div-div-div-div691-stickers-to-spell-word/0div-idbig-omega-company-tagsdiv-idbig-omega-topbardiv-classcompanytagscontainer-styleoverflow-x-scroll-flex-wrap-nowrap-div-classcompanytagscontainer-tag-stylebackground-color-rgba0-10-32-005-divfacebook-divdiv-classcompanytagscontainer-tagoccurence3-div-div-divdiv-classcompanytagscontainer-chevrondivsvg-version11-idicon-xmlnshttp-wwww3org-2000-svg-xmlns-xlinkhttp-wwww3org-1999-xlink-x0px-y0px-viewbox0-0-32-32-fill4087f1-xml-spacepreserve-stylewidth-20px-polygon-points16-22-6-12-74-106-16-192-246-106-26-12-polygonrect-id-x3c-transparent-rectangle-x3e-classst0-fillnone-width32-height32-rect-svg-div-div-div-div691-stickers-to-spell-word.cpp
class Solution {
public:
    vector<string> a;
    string t;
    int n,m,dp[1<<15];
    
    
    int recur(int mask){
        if (mask ==(1<<m)-1){
            return 0;
        }
        if (dp[mask]!=-1){return dp[mask];
                         }
        int ans = 1e9;
        for (int i=0;i<n;i++){
            vector<int> cnt(26,0);
            for (int j=0;j<a[i].size();j++){
                cnt[a[i][j]-'a']++;
            }
            int newmask =0;
            for (int j=0;j<m;j++){
                if (!(mask&(1<<j))){
                    if (cnt[t[j]-'a']>0){
                        cnt[t[j]-'a']--;
                        newmask |= (1<<j);
                    }
                }

                }
                if (newmask!=0){
                    ans = min(ans,1+recur(newmask|mask));
                
            }

        }           
        return dp[mask]=ans;

        
    }
    
    int minStickers(vector<string>& stickers, string target) {
        
        a = stickers;
        n = a.size();
        t = target;
        m= t.size();
        memset(dp,-1,sizeof(dp));
        
        int val = recur(0);
        
        if (val>=1e9){
            val=-1;
        }
        return val;
        
    }
};

File: 0div-idbig-omega-company-tagsdiv-idbig-omega-topbardiv-classcompanytagscontainer-styleoverflow-x-scroll-flex-wrap-nowrap-div-classcompanytagscontainer-tag-stylebackground-color-rgba0-10-32-005-divfacebook-divdiv-classcompanytagscontainer-tagoccurence3-div-div-divdiv-classcompanytagscontainer-chevrondivsvg-version11-idicon-xmlnshttp-wwww3org-2000-svg-xmlns-xlinkhttp-wwww3org-1999-xlink-x0px-y0px-viewbox0-0-32-32-fill4087f1-xml-spacepreserve-stylewidth-20px-polygon-points16-22-6-12-74-106-16-192-246-106-26-12-polygonrect-id-x3c-transparent-rectangle-x3e-classst0-fillnone-width32-height32-rect-svg-div-div-div-div691-stickers-to-spell-word/0div-idbig-omega-company-tagsdiv-idbig-omega-topbardiv-classcompanytagscontainer-styleoverflow-x-scroll-flex-wrap-nowrap-div-classcompanytagscontainer-tag-stylebackground-color-rgba0-10-32-005-divfacebook-divdiv-classcompanytagscontainer-tagoccurence3-div-div-divdiv-classcompanytagscontainer-chevrondivsvg-version11-idicon-xmlnshttp-wwww3org-2000-svg-xmlns-xlinkhttp-wwww3org-1999-xlink-x0px-y0px-viewbox0-0-32-32-fill4087f1-xml-spacepreserve-stylewidth-20px-polygon-points16-22-6-12-74-106-16-192-246-106-26-12-polygonrect-id-x3c-transparent-rectangle-x3e-classst0-fillnone-width32-height32-rect-svg-div-div-div-div691-stickers-to-spell-word.py
class Solution:
    def minStickers(self, stickers: List[str], target: str) -> int:
        
        m = len(target)
        mx = (1<<m)-1
        n= len(stickers)
        
        dp = [-1 for _ in range(mx+10)]
            
        def recur(mask):
            if mask == (mx):
                return 0
            
            if dp[mask]!=-1:
                return dp[mask]
            
            ans = int(1e19)
            
            for i in range(n):
                cnt = [0]*(26)
                for j in range(len(stickers[i])):
                    cnt[ord(stickers[i][j])-ord('a')]+=1
                # print(i,cnt)
                newmask =0
                
                for j in range(m):
                    if ((1<<j)&(mask))==0:
                        if cnt[ord(target[j])-ord('a')]>0:
                            cnt[ord(target[j])-ord('a')]-=1
                            newmask |= (1<<j)
                if newmask!=0:
                    ans = min(ans,1+recur(mask|newmask))
            dp[mask] = ans
            return dp[mask]
        
        val = recur(0)
        if val == int(1e19):
            return -1
        return val
                        
            
            
            